apps/ejabberd/src/rdbms/mongoose_rdbms_pgsql.erl

```diff
@@ -61,28 +76,29 @@ is_error_duplicate(_Reason) -> false.
 db_opts({pgsql, Server, DB, User, Pass}) ->
     db_opts({pgsql, Server, ?PGSQL_PORT, DB, User, Pass});
 db_opts({pgsql, Server, Port, DB, User, Pass}) when is_integer(Port) ->
-    [Server, Port, DB, User, Pass].
+    [
+     {host, Server},
+     {port, Port},
+     {database, DB},
+     {username, User},
+     {password, Pass}
+    ].
 
--spec pgsql_to_odbc(pgsql_connection:result_tuple() | {error, any()}) ->
-                           {error, any()} | {selected, [tuple()]} |
-                           {updated, undefined | non_neg_integer()}.
+-spec pgsql_to_odbc(epgsql:reply(term())) -> mongoose_rdbms:query_result().
 pgsql_to_odbc(Items) when is_list(Items) ->
     lists:reverse([pgsql_to_odbc(Item) || Item <- Items]);
 pgsql_to_odbc({error, Reason}) ->
-    {pgsql_error, Details} = Reason,
-    Message = proplists:get_value(message, Details, <<"unknown error">>),
-    {error, unicode:characters_to_list(Message)};
-pgsql_to_odbc({{select, _Count}, Rows}) ->
-    {selected, [parse_row(Row) || Row <- Rows]};
-pgsql_to_odbc({{insert, _TableOID, Count}, _Rows}) ->
-    {updated, Count};
-pgsql_to_odbc({{_, Count}, _}) when is_integer(Count) ->
+    {error, unicode:characters_to_list(element(5, Reason))};
```

Comment:
I don't like this `element(5, Reason)` here. The error tuple is a well-defined error record (check `epgsql.hrl`).

What is more, I think you could match for an error code for duplicate key and then just return something like `{error, duplicate_primary_key}` or similar. This would eliminate a need for awkward `is_error_duplicate` function.